### PR TITLE
feat: input pill list component

### PR DESCRIPTION
**Pull request by arjunlalb**
## Description
A new component to support inputting values into a list of pills.
Takes a list of values as input and renders them as pills.

Future improvements:
1. Add validation
2. Add support for enabling/disabling user input.

### Testing
Tested on local. UTs added.

### Checklist:
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

Screenshot 2022-12-19 at 1 34 18 PM

https://user-images.githubusercontent.com/63222211/208375508-2155d07d-6e69-44cf-b00a-901b6899bc0c.mov

### CODE REVIEW

File: projects/common/src/key-codes/key-code.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export const enum KeyCode {
+  Enter = 'Enter',
+  Comma = ','
+}
+
+export const isEnterOrCommaKeyEvent = (event: KeyboardEvent): boolean =>
```
Comment (adisreyaj) [inline projects/common/src/key-codes/key-code.ts:6]: Should we make this configurable? Enabled `Enter to add new` or `Comma to add new`?

Comment (arjunlalb) [reply projects/common/src/key-codes/key-code.ts:6]: will add a general method in a follow up.

File: projects/components/src/input-pill-list/input-pill-list.component.test.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    });
+    spectator.triggerEventHandler(secondPill, 'valueChange', undefined);
+
+    // Debounce of 200ms is applied on the pill input update. Wait for it.
+    spectator.tick(250);
+    expect(getPillValues(spectator)).toEqual(['input-from-buffer', 'i-am-test-2-now', 'test-2']);
+    expect(valueChangeEmitterSpy).toHaveBeenCalledWith(['input-from-buffer', 'i-am-test-2-now', 'test-2']);
+    expect(valueChangeEmitterSpy).toHaveBeenCalledTimes(2);
+  }));
+});
```
[the remark this answers is not in the document]
Comment (arjunlalb) [reply projects/components/src/input-pill-list/input-pill-list.component.test.ts:97]: some of the checks are looking into form controls along with the emissions.

File: projects/components/src/input-pill-list/input-pill-list.component.scss [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    align-items: center;
+    width: 100%;
+    line-height: $input-height;
+    border-radius: 8px;
+    display: block;
+  }
+
+  .pill-list {
+    display: grid;
+    grid-template-columns: 1fr 1fr;
```
[the remark this answers is not in the document]
Comment (arjunlalb) [reply projects/components/src/input-pill-list/input-pill-list.component.scss:33]: perhaps. our current usecase is for 2 per row grid. will keep this in mind.

File: projects/components/src/input-pill-list/input-pill-list.component.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+import { InputAppearance } from '../input/input-appearance';
+
+@Component({
+  selector: 'ht-input-pill-list',
+  template: `
+    <div class="input-pill-list" [formGroup]="this.form">
+      <div class="header">
+        <ht-form-field class="form-field">
+          <ht-input
+            [disabled]="this.disabled || this.disableAdd || this.disableUpdate"
```
Comment (adisreyaj) [inline projects/components/src/input-pill-list/input-pill-list.component.ts:24]: nit:
```suggestion
            [disabled]="this.shouldDisableUpdate || this.disableAdd"
```

File: projects/components/src/input-pill-list/input-pill-list.component.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+  @Input()
+  public readonly disableUpdate: boolean = false;
+
+  @Input()
+  public readonly disabled: boolean = false;
+
+  @Output()
+  public readonly valueChange: EventEmitter<string[]> = new EventEmitter();
+
+  public form: FormGroup = new FormGroup({
```
Comment (adisreyaj) [inline projects/components/src/input-pill-list/input-pill-list.component.ts:84]:
```suggestion
  public readonly form: FormGroup = new FormGroup({
```

**Review by adisreyaj — APPROVED** [review]

File: projects/components/src/input-pill-list/input-pill-list.component.scss [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    .pill {
+      @include m.body-small(m.$gray-9);
+      align-items: center;
+      background-color: m.$blue-1;
+      display: flex;
+      justify-content: space-between;
+      gap: 4px;
+      border: 1px solid m.$gray-5;
+      border-radius: $border-radius;
+      padding: 2px 4px 2px;
```
Comment (itssharmasandeep) [inline projects/components/src/input-pill-list/input-pill-list.component.scss:50]:
```suggestion
      padding: 2px 4px;
```

File: projects/components/src/input-pill-list/input-pill-list.component.ts [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+  @Input()
+  public readonly disableUpdate: boolean = false;
+
+  @Input()
+  public readonly disabled: boolean = false;
+
+  @Output()
+  public readonly valueChange: EventEmitter<string[]> = new EventEmitter();
+
+  public readonly form: FormGroup = new FormGroup({
```
Comment (itssharmasandeep) [inline projects/components/src/input-pill-list/input-pill-list.component.ts:84]: This is implementing a `ControlValueAccessor ` and it also has form-group used

File: projects/components/src/input-pill-list/input-pill-list.component.ts [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+})
+export class InputPillListComponent implements ControlValueAccessor, OnChanges {
+  @Input()
+  public readonly values: string[] = [];
+
+  @Input()
+  public readonly disableAdd: boolean = false;
+
+  @Input()
+  public readonly disableUpdate: boolean = false;
```
Comment (itssharmasandeep) [inline projects/components/src/input-pill-list/input-pill-list.component.ts:76]: is it only used for removing the pill?

File: projects/components/src/input-pill-list/input-pill-list.component.ts (hunk 7 above)
Comment (arjunlalb) [reply projects/components/src/input-pill-list/input-pill-list.component.ts:84]: Yes. Internally this component has a form based implementation. ControlValueAccessor allows for this component to be used within forms as well.

File: projects/components/src/input-pill-list/input-pill-list.component.ts (hunk 8 above)
Comment (arjunlalb) [reply projects/components/src/input-pill-list/input-pill-list.component.ts:76]: also to control editing the values of pills.

**Review by itssharmasandeep — APPROVED** [review]

**Pull request merged**